plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -0,0 +1,482 @@
+import java.util.ArrayList;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.charset.Charset;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.stream.Stream;
+import java.util.Iterator;
+import java.nio.file.PathMatcher;
+
+import org.apache.commons.io.filefilter.WildcardFileFilter;
+import org.apache.commons.io.IOUtils;
+
+import java.io.FileFilter;
+
+import org.apache.commons.io.FilenameUtils;
+
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.jar.Attributes;
+import java.util.Map;
+import java.util.HashMap;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.io.FileUtils;
+
+import java.net.URL;
+
+import org.json.JSONObject;
+import org.json.JSONArray;
+
+import io.jenkins.tools.pluginmanager.config.Config;
+
+
+public class PluginManager {
+    private List<Plugin> plugins;
+    private List<Plugin> failedPlugins;
+    private File refDir;
+
+    private String JENKINS_UC_LATEST = "";
+    private final String JENKINS_UC_EXPERIMENTAL = "https://updates.jenkins.io/experimental";
+    private final String JENKINS_INCREMENTALS_REPO_MIRROR = "https://repo.jenkins-ci.org/incrementals";
+    private final String JENKINS_UC = "https://updates.jenkins.io";
+    private final String JENKINS_UC_DOWNLOAD = JENKINS_UC + "/download";
+    private final String JENKINS_UC_JSON = "https://updates.jenkins.io/update-center.json";
+    private final String SEPARATOR = File.separator;
+
+    //private final String JENKINS_WAR = "/usr/share/jenkins/jenkins.war";
+    public static final String JENKINS_WAR = "/Users/natashastopa/Jenkins/Jenkins/test2/mywar-1.0-SNAPSHOT.war";
+
+    private String jenkinsVersion;
+
+    private File jenkinsWarFile;
+    private Map<String, String> installedPluginVersions;
+    private Map<String, String> bundledPluginVersions;
+    Config cfg;
+
+
+    public PluginManager(Config cfg) {
+        plugins = new ArrayList<>();
+        failedPlugins = new ArrayList();
+        jenkinsWarFile = new File(JENKINS_WAR);
+        installedPluginVersions = new HashMap<>();
+        bundledPluginVersions = new HashMap<>();
+        refDir = cfg.getPluginDir();
+        this.cfg = cfg;
+    }
+
+
+    public void start() {
+        if (!refDir.mkdir()) {
+            System.out.println("Unable to create plugin directory");
+        }
+
+        jenkinsVersion = getJenkinsVersion();
+
+        //check if version specific update center
+        if (!StringUtils.isEmpty(jenkinsVersion)) {
+            JENKINS_UC_LATEST = new StringBuilder("https://updates.jenkins.io/").append(jenkinsVersion).toString();
+            try {
+                URL url = new URL(JENKINS_UC_LATEST);
+                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
```

Comment:
It would probably be simpler to use a more modern http client,
You're going to have issues with things like redirects, proxies, ssl and reading the response is not that straightforward.

Some clients in order of ease of use:
okhttp,
unirest
apache http client
httpurl connection

java 11+
java http client

okhttp and unirest are both probably just as easy to use...

Reply:
Fixed in 8d0cc18dd262d98f14e5c81c85ffade797a1df0d